File: ls25-backend/app/calculations.py
from typing import List, Dict, Set
from sqlalchemy.orm import Session
from app.database import Ware, Produktion, ProductionInput, ProductionOutput
from app.schemas import ChainCalculationResult, RevenueCalculationResult, CapacityAnalysisResult

def calculate_chain(db: Session, start_good_id: str, quantity: float) -> List[ChainCalculationResult]:
    """
    Kettenrechner: Calculate what can be produced from a starting good and quantity
    """
    results = []
    visited_productions: Set[str] = set()
    
    def process_good(good_id: str, available_quantity: float, stage: int):
        good = db.query(Ware).filter(Ware.id == good_id).first()
        if good:
            results.append(ChainCalculationResult(
                good_id=good_id,
                good_name=good.name,
                quantity=available_quantity,
                stage=stage
            ))
        
        productions = db.query(Produktion).join(ProductionInput).filter(
            ProductionInput.good_id == good_id
        ).all()
        
        for production in productions:
            if production.id in visited_productions:
                continue
            visited_productions.add(production.id)
            
            input_req = db.query(ProductionInput).filter(
                ProductionInput.production_id == production.id,
                ProductionInput.good_id == good_id
            ).first()
            
            if input_req and input_req.quantity_per_cycle > 0:
                max_cycles = available_quantity / input_req.quantity_per_cycle
                
                outputs = db.query(ProductionOutput).filter(
                    ProductionOutput.production_id == production.id
                ).all()
                
                for output in outputs:
                    output_quantity = max_cycles * output.quantity_per_cycle
                    process_good(output.good_id, output_quantity, stage + 1)
    
    process_good(start_good_id, quantity, 0)
    return results

def calculate_revenue(db: Session, production_ids: List[str], runtime_months: float = 1.0) -> RevenueCalculationResult:
    """
    Erlösrechnung: Calculate gross and net revenue for production chain
    """
    gross_revenue = 0.0
    production_costs = 0.0
    breakdown = {
        "productions": [],
        "total_fixed_costs": 0.0,
        "total_variable_costs": 0.0,
        "total_building_costs": 0.0
    }
    
    for production_id in production_ids:
        production = db.query(Produktion).filter(Produktion.id == production_id).first()
        if not production:
            continue
        
        cycles_run = production.cycles_per_month * runtime_months
        
        fixed_costs = production.fixed_costs_per_month * runtime_months
        variable_costs = production.variable_costs_per_cycle * cycles_run
        
        building_costs = 0.0
        for building in production.buildings:
            building_costs += building.building_costs_per_month * runtime_months
        
        production_costs += fixed_costs + variable_costs + building_costs
        
        for output in production.outputs:
            output_quantity = cycles_run * output.quantity_per_cycle
            price_per_unit = output.good.price_per_1000l / 1000.0
            revenue = output_quantity * price_per_unit
            gross_revenue += revenue
        
        breakdown["productions"].append({
            "id": production_id,
            "name": production.name,
            "cycles_run": cycles_run,
            "fixed_costs": fixed_costs,
            "variable_costs": variable_costs,
            "building_costs": building_costs,
            "revenue": revenue
        })
        
        breakdown["total_fixed_costs"] += fixed_costs
        breakdown["total_variable_costs"] += variable_costs
        breakdown["total_building_costs"] += building_costs
    
    net_revenue = gross_revenue - production_costs
    
    return RevenueCalculationResult(
        gross_revenue=gross_revenue,
        production_costs=production_costs,
        net_revenue=net_revenue,
        breakdown=breakdown
    )

def calculate_capacity_analysis(db: Session, production_id: str) -> CapacityAnalysisResult:
    """
    Auslastungsanalyse: Calculate required inputs for 100% capacity utilization
    """
    production = db.query(Produktion).filter(Produktion.id == production_id).first()
    if not production:
        raise ValueError("Production not found")
    
    inputs_per_month = []
    inputs_per_year = []
    
    for input_item in production.inputs:
        monthly_requirement = input_item.quantity_per_cycle * production.cycles_per_month
        yearly_requirement = monthly_requirement * 12
        
        inputs_per_month.append({
            "good_id": input_item.good_id,
            "good_name": input_item.good.name,
            "quantity": monthly_requirement,
            "unit": input_item.good.unit
        })
        
        inputs_per_year.append({
            "good_id": input_item.good_id,
            "good_name": input_item.good.name,
            "quantity": yearly_requirement,
            "unit": input_item.good.unit
        })
    
    return CapacityAnalysisResult(
        production_id=production_id,
        production_name=production.name,
        inputs_per_month=inputs_per_month,
        inputs_per_year=inputs_per_year
    )

def calculate_processing_time(cycles_per_month: float) -> dict:
    """
    Calculate processing time based on cycles per month
    """
    if cycles_per_month <= 0:
        return {"seconds_per_cycle": 0, "minutes_per_cycle": 0, "hours_per_cycle": 0}
    
    seconds_per_month = 30 * 24 * 3600
    seconds_per_cycle = seconds_per_month / cycles_per_month
    
    return {
        "seconds_per_cycle": seconds_per_cycle,
        "minutes_per_cycle": seconds_per_cycle / 60,
        "hours_per_cycle": seconds_per_cycle / 3600
    }

File: ls25-backend/src/calculations.ts
import { db } from './database';
import { ChainCalculationResult, RevenueCalculationResult, CapacityAnalysisResult } from './types';

export function calculateChain(startGoodId: string, quantity: number): ChainCalculationResult[] {
  const results: ChainCalculationResult[] = [];
  const visitedProductions = new Set<string>();

  function processGood(goodId: string, availableQuantity: number, stage: number) {
    const good = db.getWareById(goodId);
    if (good) {
      results.push({
        good_id: goodId,
        good_name: good.name,
        quantity: availableQuantity,
        stage
      });
    }

    const produktionen = db.getAllProduktionen().filter(p => 
      p.inputs.some(input => input.good_id === goodId)
    );

    for (const produktion of produktionen) {
      if (visitedProductions.has(produktion.id)) {
        continue;
      }
      visitedProductions.add(produktion.id);

      const inputReq = produktion.inputs.find(input => input.good_id === goodId);
      if (inputReq && inputReq.quantity_per_cycle > 0) {
        const maxCycles = availableQuantity / inputReq.quantity_per_cycle;

        for (const output of produktion.outputs) {
          const outputQuantity = maxCycles * output.quantity_per_cycle;
          processGood(output.good_id, outputQuantity, stage + 1);
        }
      }
    }
  }

  processGood(startGoodId, quantity, 0);
  return results;
}

export function calculateRevenue(productionIds: string[], runtimeMonths: number = 1.0): RevenueCalculationResult {
  let grossRevenue = 0.0;
  let productionCosts = 0.0;
  const breakdown = {
    productions: [] as any[],
    total_fixed_costs: 0.0,
    total_variable_costs: 0.0,
    total_building_costs: 0.0
  };

  for (const productionId of productionIds) {
    const produktion = db.getProduktionById(productionId);
    if (!produktion) continue;

    const cyclesRun = produktion.cycles_per_month * runtimeMonths;
    const fixedCosts = produktion.fixed_costs_per_month * runtimeMonths;
    const variableCosts = produktion.variable_costs_per_cycle * cyclesRun;

    let buildingCosts = 0.0;
    const gebaeude = db.getAllGebaeude().filter(g => 
      g.productions.some(p => p.id === productionId)
    );
    
    for (const building of gebaeude) {
      buildingCosts += building.building_costs_per_month * runtimeMonths;
    }

    productionCosts += fixedCosts + variableCosts + buildingCosts;

    let revenue = 0.0;
    for (const output of produktion.outputs) {
      const outputQuantity = cyclesRun * output.quantity_per_cycle;
      const pricePerUnit = (output.good?.price_per_1000l || 0) / 1000.0;
      revenue += outputQuantity * pricePerUnit;
    }
    grossRevenue += revenue;

    breakdown.productions.push({
      id: productionId,
      name: produktion.name,
      cycles_run: cyclesRun,
      fixed_costs: fixedCosts,
      variable_costs: variableCosts,
      building_costs: buildingCosts,
      revenue
    });

    breakdown.total_fixed_costs += fixedCosts;
    breakdown.total_variable_costs += variableCosts;
    breakdown.total_building_costs += buildingCosts;
  }

  const netRevenue = grossRevenue - productionCosts;

  return {
    gross_revenue: grossRevenue,
    production_costs: productionCosts,
    net_revenue: netRevenue,
    breakdown
  };
}

export function calculateCapacityAnalysis(productionId: string): CapacityAnalysisResult {
  const produktion = db.getProduktionById(productionId);
  if (!produktion) {
    throw new Error('Production not found');
  }

  const inputsPerMonth: Array<{
    good_id: string;
    good_name: string;
    quantity: number;
    unit: string;
  }> = [];
  const inputsPerYear: Array<{
    good_id: string;
    good_name: string;
    quantity: number;
    unit: string;
  }> = [];

  for (const inputItem of produktion.inputs) {
    const monthlyRequirement = inputItem.quantity_per_cycle * produktion.cycles_per_month;
    const yearlyRequirement = monthlyRequirement * 12;

    inputsPerMonth.push({
      good_id: inputItem.good_id,
      good_name: inputItem.good?.name || 'Unknown',
      quantity: monthlyRequirement,
      unit: inputItem.good?.unit || 'Unknown'
    });

    inputsPerYear.push({
      good_id: inputItem.good_id,
      good_name: inputItem.good?.name || 'Unknown',
      quantity: yearlyRequirement,
      unit: inputItem.good?.unit || 'Unknown'
    });
  }

  return {
    production_id: productionId,
    production_name: produktion.name,
    inputs_per_month: inputsPerMonth,
    inputs_per_year: inputsPerYear
  };
}

export function calculateProcessingTime(cyclesPerMonth: number): any {
  if (cyclesPerMonth <= 0) {
    return { seconds_per_cycle: 0, minutes_per_cycle: 0, hours_per_cycle: 0 };
  }

  const secondsPerMonth = 30 * 24 * 3600;
  const secondsPerCycle = secondsPerMonth / cyclesPerMonth;

  return {
    seconds_per_cycle: secondsPerCycle,
    minutes_per_cycle: secondsPerCycle / 60,
    hours_per_cycle: secondsPerCycle / 3600
  };
}
